File: ch07/CSharpServerControl1/Listing07-41.cs

using System;
using System.ComponentModel;
using System.ComponentModel.Design;
using System.Text;
using System.Web.UI;
using System.Web.UI.Design;
namespace CSharpServerControl1
{
public class MultiRegionControlDesigner :
    System.Web.UI.Design.WebControls.CompositeControlDesigner {
          
    protected int _currentView = 0;
          
    private Listing0740 myControl;
    public override void Initialize(IComponent component)
    {
        base.Initialize(component);
        myControl = (Listing0740)component;
    }
          
    public override bool AllowResize { get { return true;}}
          
    protected override void OnClick(DesignerRegionMouseEventArgs e)
    {
        if (e.Region == null)
            return;
          
        if (e.Region.Name == "Header0" && _currentView != 0) {
            _currentView = 0;
            UpdateDesignTimeHtml();
        }
          
        if (e.Region.Name == "Header1" && _currentView != 1) {
            _currentView = 1;
            UpdateDesignTimeHtml();
        }
    }
          
    public override String GetDesignTimeHtml(DesignerRegionCollection regions)
    {
        BuildRegions(regions);
        return BuildDesignTimeHtml();
    }
          
    protected virtual void BuildRegions(DesignerRegionCollection regions)
    {
        regions.Add(new DesignerRegion(this, "Header0"));
        regions.Add(new DesignerRegion(this, "Header1"));
          
        // If the current view is for all, we need another editable region
        EditableDesignerRegion edr0 = new
            EditableDesignerRegion(this, "Content" + _currentView, false);
        edr0.Description = "Add stuff in here if you dare:";
        regions.Add(edr0);
          
        // Set the highlight, depending upon the selected region
        if (_currentView ==0 || _currentView==1)
            regions[_currentView].Highlight = true;
    }
          
    protected virtual string BuildDesignTimeHtml()
    {
        StringBuilder sb = new StringBuilder();
        sb.Append(BuildBeginDesignTimeHtml());
        sb.Append(BuildContentDesignTimeHtml());
        sb.Append(BuildEndDesignTimeHtml());
          
        return sb.ToString();
    }
          
    protected virtual String BuildBeginDesignTimeHtml()
    {
        // Create the table layout
        StringBuilder sb = new StringBuilder();
        sb.Append("<table ");
          
        // Styles that we'll use to render for the design-surface
        sb.Append("height='" + myControl.Height.ToString() + "' width='" +
            myControl.Width.ToString() +  "'>");
          
        // Generate the title or caption bar
        sb.Append("<tr height='25px' align='center' " +
            "style='font-family:tahoma;font-size:10pt;font-weight:bold;'>" +
            "<td style='width:50%' " + DesignerRegion.DesignerRegionAttributeName +
            "='0'>");
        sb.Append("Page-View 1</td>");
        sb.Append("<td style='width:50%' " +
            DesignerRegion.DesignerRegionAttributeName + "='1'>");
        sb.Append("Page-View 2</td></tr>");
          
        return sb.ToString();
    }
          
    protected virtual String BuildEndDesignTimeHtml()
    {
        return ("</table>");
    }
          
    protected virtual String BuildContentDesignTimeHtml()
    {
        StringBuilder sb = new StringBuilder();
        sb.Append("<td colspan='2' style='");
        sb.Append("background-color:" + myControl.BackColor.Name.ToString() +
            ";' ");
          
        sb.Append(DesignerRegion.DesignerRegionAttributeName + "='2'>");
          
        return sb.ToString();
    }
          
    public override string GetEditableDesignerRegionContent
        (EditableDesignerRegion region)
    {
        IDesignerHost host =
            (IDesignerHost)Component.Site.GetService(typeof(IDesignerHost));
          
        if (host != null) {
            ITemplate template = myControl.View1;
          
            if (region.Name == "Content1")
                template = myControl.View2;
          
            if (template != null)
                return ControlPersister.PersistTemplate(template, host);
        }
          
        return String.Empty;
    }
          
    public override void SetEditableDesignerRegionContent
         (EditableDesignerRegion region, string content)
    {
        int regionIndex = Int32.Parse(region.Name.Substring(7));
          
        if (content == null)
        {
            if (regionIndex == 0)
                myControl.View1 = null;
            else if (regionIndex == 1)
                myControl.View2 = null;
            return;
        }
          
        IDesignerHost host =
            (IDesignerHost)Component.Site.GetService(typeof(IDesignerHost));
          
        if (host != null)
        {
            ITemplate template = ControlParser.ParseTemplate(host, content);
          
            if (template != null)
            {
                if (regionIndex == 0)
                    myControl.View1 = template;
                else if (regionIndex == 1)
                    myControl.View2 = template;
            }
        }
    }
}

}

File: ch07/VbServerControl1/Listing07-41.vb
Imports System.ComponentModel
Imports System.Web.UI.Design
Imports System.Text
Imports System.ComponentModel.Design
Imports System.Web.UI

Public Class MultiRegionControlDesigner
    Inherits System.Web.UI.Design.WebControls.CompositeControlDesigner

    Protected _currentView As Int32 = 0
    Private myControl As Listing0740

    Public Overrides Sub Initialize(ByVal component As IComponent)
        MyBase.Initialize(component)
        myControl = CType(component, Listing0740)
    End Sub

    Public Overrides ReadOnly Property AllowResize() As Boolean
        Get
            Return True
        End Get
    End Property

    Protected Overrides Sub OnClick(ByVal e As DesignerRegionMouseEventArgs)

        If (e.Region Is Nothing) Then
            Return
        End If

        If ((e.Region.Name = "Header0") And (Not _currentView = 0)) Then
            _currentView = 0
            UpdateDesignTimeHtml()
        End If

        If ((e.Region.Name = "Header1") And (Not _currentView = 1)) Then

            _currentView = 1
            UpdateDesignTimeHtml()
        End If
    End Sub

    Public Overrides Function GetDesignTimeHtml( _
            ByVal regions As DesignerRegionCollection) As String
        BuildRegions(regions)
        Return BuildDesignTimeHtml()
    End Function

    Protected Overridable Sub BuildRegions( _
            ByVal regions As DesignerRegionCollection)

        regions.Add(New DesignerRegion(Me, "Header0"))
        regions.Add(New DesignerRegion(Me, "Header1"))

        ' If the current view is for all, we need another editable region
        Dim edr0 As New EditableDesignerRegion(Me, "Content" & _currentView, False)
        edr0.Description = "Add stuff in here if you dare:"
        regions.Add(edr0)

        ' Set the highlight, depending upon the selected region
        If ((_currentView = 0) Or (_currentView = 1)) Then
            regions(_currentView).Highlight = True
        End If
    End Sub

    Protected Overridable Function BuildDesignTimeHtml() As String

        Dim sb As New StringBuilder()
        sb.Append(BuildBeginDesignTimeHtml())
        sb.Append(BuildContentDesignTimeHtml())
        sb.Append(BuildEndDesignTimeHtml())

        Return sb.ToString()
    End Function

    Protected Overridable Function BuildBeginDesignTimeHtml() As String
        ' Create the table layout
        Dim sb As New StringBuilder()
        sb.Append("<table ")

        ' Styles that we'll use to render for the design-surface
        sb.Append("height='" & myControl.Height.ToString() & "' width='" & _
            myControl.Width.ToString() & "'>")

        ' Generate the title or caption bar
        sb.Append("<tr height='25px' align='center' " & _
            "style='font-family:tahoma;font-size:10pt;font-weight:bold;'>" & _
            "<td style='width:50%' " & _
            DesignerRegion.DesignerRegionAttributeName & "='0'>")
        sb.Append("Page-View 1</td>")
        sb.Append("<td style='width:50%' " & _
            DesignerRegion.DesignerRegionAttributeName & "='1'>")
        sb.Append("Page-View 2</td></tr>")

        Return sb.ToString()
    End Function

    Protected Overridable Function BuildEndDesignTimeHtml() As String
        Return ("</table>")
    End Function

    Protected Overridable Function BuildContentDesignTimeHtml() As String

        Dim sb As New StringBuilder()
        sb.Append("<td colspan='2' style='")
        sb.Append("background-color:" & _
            myControl.BackColor.Name.ToString() & ";' ")

        sb.Append(DesignerRegion.DesignerRegionAttributeName & "='2'>")

        Return sb.ToString()
    End Function

    Public Overrides Function GetEditableDesignerRegionContent( _
            ByVal region As EditableDesignerRegion) As String

        Dim host As IDesignerHost = _
            CType(Component.Site.GetService(GetType(IDesignerHost)), IDesignerHost)

        If (Not host Is Nothing) Then
            Dim template As ITemplate = myControl.View1

            If (region.Name = "Content1") Then
                template = myControl.View2
            End If

            If (Not template Is Nothing) Then
                Return ControlPersister.PersistTemplate(template, host)
            End If

        End If

        Return String.Empty
    End Function

    Public Overrides Sub SetEditableDesignerRegionContent( _
            ByVal region As EditableDesignerRegion, ByVal content As String)

        Dim regionIndex As Int32 = Int32.Parse(region.Name.Substring(7))

        If (content Is Nothing) Then

            If (regionIndex = 0) Then
                myControl.View1 = Nothing
            ElseIf (regionIndex = 1) Then
                myControl.View2 = Nothing
                Return
            End If

            Dim host As IDesignerHost = _
                CType(Component.Site.GetService(GetType(IDesignerHost)), 
                    IDesignerHost)

            If (Not host Is Nothing) Then
                Dim template = ControlParser.ParseTemplate(host, content)

                If (Not template Is Nothing) Then
                    If (regionIndex = 0) Then
                        myControl.View1 = template
                    ElseIf (regionIndex = 1) Then
                        myControl.View2 = template
                    End If
                End If
            End If
        End If
    End Sub
End Class